File: cpp/logger/logger.cpp

#include "logger.h"
#include <android/log.h>

namespace bedrock::logger {

void Logger::init() {
    __android_log_print(ANDROID_LOG_INFO, LOG_TAG, "Logger initialized - سجلات Native باستخدام Android Log");
}

void Logger::log(int level, const std::string& tag, const std::string& msg) {
    android_LogPriority prio = ANDROID_LOG_INFO;
    if (level == 0) prio = ANDROID_LOG_DEBUG;
    else if (level == 2) prio = ANDROID_LOG_ERROR;
    __android_log_print(prio, tag.c_str(), "%s", msg.c_str());
}

void Logger::i(const std::string& tag, const std::string& msg) {
    __android_log_print(ANDROID_LOG_INFO, tag.c_str(), "%s", msg.c_str());
}

void Logger::e(const std::string& tag, const std::string& msg) {
    __android_log_print(ANDROID_LOG_ERROR, tag.c_str(), "%s", msg.c_str());
}

} // namespace

File: app/src/main/java/com/bedrock/client/logger/Logger.kt

package com.bedrock.client.logger
import android.content.Context
import android.util.Log
import timber.log.Timber
import java.io.File

object Logger {
    fun init(context: Context) {
        if (Timber.forest().isEmpty()) {
            Timber.plant(Timber.DebugTree())
            Timber.plant(FileTree(File(context.filesDir, "logs")))
        }
    }
    fun i(tag: String, msg: String) = Timber.tag(tag).i(msg)
    fun e(tag: String, msg: String, e: Throwable? = null) = Timber.tag(tag).e(e, msg)
    fun d(tag: String, msg: String) = Timber.tag(tag).d(msg)
    fun nativeLog(level: Int, tag: String, msg: String) {
        when(level) { 0 -> d(tag, msg); 1 -> i(tag, msg); 2 -> e(tag, msg) }
    }

    class FileTree(private val dir: File) : Timber.Tree() {
        init { dir.mkdirs() }
        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
            try {
                val f = File(dir, "client.log")
                f.appendText("${System.currentTimeMillis()} $tag: $message\n")
                if (t != null) f.appendText(t.stackTraceToString() + "\n")
            } catch (_: Exception) {}
        }
    }
}
